swift/ql/lib/codeql/swift/regex/internal/ParseRegex.qll

```diff
@@ -274,6 +274,58 @@ abstract class RegExp extends Expr {
 
   private predicate isGroupStart(int i) { this.nonEscapedCharAt(i) = "(" and not this.inCharSet(i) }
 
+  /**
+   * Holds if a parse mode starts between `start` and `end`.
+   */
+  private predicate flagGroupStart(int start, int end) {
+    this.isGroupStart(start) and
+    this.getChar(start + 1) = "?" and
+    this.getChar(start + 2) in ["i", "x", "s", "m", "w"] and
+    end = start + 2
+  }
+
+  /**
+   * Holds if a parse mode group is between `start` and `end`, and includes the
+   * mode flag `c`. For example the following span, with mode flag `i`:
+   * ```
+   * (?i)
+   * ```
+   */
+  private predicate flagGroup(int start, int end, string c) {
+    exists(int inStart, int inEnd |
+      this.flagGroupStart(start, inStart) and
+      this.groupContents(start, end, inStart, inEnd) and
+      this.getChar([inStart .. inEnd - 1]) = c
+    )
+  }
+
+  /**
+   * Gets a mode of this regular expression string if it is defined by a mode prefix.
+   */
+  string getModeFromPrefix() {
+    exists(string c | this.flagGroup(_, _, c) |
+      c = "i" and result = "IGNORECASE" // case insensitive
+      or
+      c = "x" and result = "VERBOSE" // ignores whitespace and `#` comments within patterns
+      or
+      c = "s" and result = "DOTALL" // dot matches all characters, including line terminators
+      or
+      c = "m" and result = "MULTILINE" // `^` and `$` also match beginning and end of lines
+      or
+      c = "w" and result = "UNICODE" // Unicode UAX 29 word boundary mode
+    )
+  }
+
+  /**
+   * Gets a mode (if any) of this regular expression. Can be any of:
+   * IGNORECASE
+   * VERBOSE
+   * DOTALL
+   * MULTILINE
+   * UNICODE
+   */
+  string getAMode() { result = this.getModeFromPrefix() }
```

Comment:
What's the advantage of splitting the implementation out into the separate `getModeFromPrefix`?

Reply:
`getModeFromPrefix` gets mode flags set within the regular expression string itself, i.e. the `i` in `(?i)foo`.

`getAMode` gets any mode flags that can affect this regular expression, whether they be set inside the regular expression string or programmatically applied (e.g. due to a call to [Regex.ignoresCase(_:)](https://developer.apple.com/documentation/swift/regex/ignorescase(_:)) at some point before the regex is evaluated).  As yet we only support the former, but you can expect `getAMode` to do more in future PRs.